File: Department of mathematics/2_Vector_Autoregression.py
import pandas as pd
import numpy as np
from statsmodels.tsa.stattools import adfuller
from statsmodels.tsa.vector_ar.var_model import VAR
from statsmodels.stats.outliers_influence import variance_inflation_factor

# Load the dataset
df = pd.read_csv('time_series.csv')

# Drop the 'Year' column
df = df.drop(columns=['Year'])

# Drop constant columns
df = df.loc[:, (df != df.iloc[0]).any()]

# Perform stationary testing and store the stationary variables
stationary_columns = []
for column in df.columns:
    result = adfuller(df[column])
    if result[1] < 0.05:  # if p-value < 0.05, the series is stationary
        stationary_columns.append(column)

# Perform Vector autoregression on the stationary columns three by three
results = []
for i in range(0, len(stationary_columns), 3):
    try:
        model = VAR(df[stationary_columns[i:i+3]])
        model_fit = model.fit()
        results.append(model_fit.summary())
    except np.linalg.LinAlgError:
        print(f"Skipping VAR for columns {stationary_columns[i:i+3]} due to LinAlgError")

# Save the results in a new csv file
with open('var_results.csv', 'w') as f:
    for result in results:
        f.write(str(result))

# Calculate Variance Inflation Factor for each variable
vif_data = pd.DataFrame()
vif_data["feature"] = df.columns
vif_data["VIF"] = [variance_inflation_factor(df.values, i) for i in range(len(df.columns))]

# Print the VIF results
print(vif_data)

File: Department of Media and Information Technology/2_Vector_Autoregression.R
library(vars)
library(car)

# Load the dataset
df <- read.csv("time_series.csv")

# Drop the 'Year' column
df$Year <- NULL

# Drop constant columns
df <- df[, sapply(df, function(x) length(unique(x)) > 1)]

# Perform stationary testing and store the stationary variables
for (column in colnames(df)) {
  result <- ur.df(df[[column]], type = "drift")
  test_statistic <- summary(result)@teststat[1, "tau2"]
  critical_value_5pct <- summary(result)@cval[1, "5pct"]
  if (test_statistic < critical_value_5pct) {  # if test statistic is less than the 5% critical value, the series is stationary
    stationary_columns <- c(stationary_columns, column)
  }
}

# Perform Vector autoregression on the stationary columns two by two
results <- list()
for (i in seq(1, length(stationary_columns), by = 2)) {
  tryCatch({
    model <- VAR(df[, stationary_columns[i:(i+1)]], lag.max = 4, type = "const")
    results[[length(results) + 1]] <- summary(model)
  }, error = function(e) {
    print(paste("Skipping VAR for columns", stationary_columns[i:(i+1)], "due to LinAlgError"))
  })
}

# Save the results in a new csv file
write.table(do.call(rbind, results), file = "var_results.csv", sep = ";", row.names = FALSE)

# Remove rows with missing values
df_complete <- df[complete.cases(df), ]

# Check if the number of observations is greater than the number of variables
if (nrow(df_complete) > ncol(df_complete)) {
  vif_data <- data.frame(feature = colnames(df_complete))
  vif_data$VIF <- vif(lm(paste0("~ ."), data = df_complete))
  print(vif_data)
} else {
  print("Cannot calculate VIF: Too many variables compared to the number of observations.")
}
